File: leetcode/leetcode-42.py
from typing import List

class Solution:
    def trap(self, height: List[int]) -> int:
        ans = 0
        le = 0
        re = len(height)-1
        lmax = 0
        rmax = 0
        while le < re:
            if height[le] <= height[re]:
                if height[le] >= lmax:
                    lmax = height[le]
                else:
                    ans += lmax - height[le]
                le += 1
            else:
                if height[re] >= rmax:
                    rmax = height[re]
                else:
                    ans += rmax - height[re]
                re -= 1
        return ans


if __name__ == "__main__":
    so = Solution()
    print(so.trap([0,1,0,2,1,0,1,3,2,1,2,1]))

File: leetcode/leetcode-42.cpp
/*
 *@File Name          :leetcode-42.cpp
 *@Author             :cc
 *@Version            :1.0.0
 *@Date               :2022/2/28 9:53
 *@Description        :
 *@Function List      :
 *@History            :
 */

#include "algorithm"
#include "vector"
#include "string"

using namespace std;

class Solution {
public:
    int trap(vector<int> &height) {
        int n = height.size();
        vector<int> preMax(n, INT_MIN);
        for (int i = 1; i < n; i++)
            preMax[i] = max(preMax[i - 1], height[i - 1]);
        int postMax = INT_MIN;
        int ans = 0;
        for (int i = n - 2; i >= 0; i--) {
            postMax = max(postMax, height[i + 1]);
            ans += min(min(postMax, preMax[i]) - height[i], 0);
        }
        return ans;
    }
};